### PR TITLE
Implemented init command in cli

### PR DESCRIPTION
**Pull request by gaborschulz**
*Description of changes:*
Implemented init command in the CLI

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cid/commands/cmd_init.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        # Create table with predefined fields
+        self._create_glue_crawler_role()
+        self._attach_policies_to_crawler_role()
+
+        self._create_glue_crawler()
+        # self._create_glue_table()
+        # Create crawler that updates the table
+
+    def _get_base_data(self):
+        self.aws_partition = get_parameter("aws-partition", "Please, choose an AWS partition", choices=["aws", "aws-cn"], default="aws")
```
Comment (iakov-aws) [inline cid/commands/cmd_init.py:57]: not needed. you can get it from the current region

```
import boto3

session = boto3.Session()
for partition in session.get_available_partitions():
    for region in session.get_available_regions('quicksight', partition):
        print(region, partition)
```

File: cid/commands/cmd_init.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    unset_parameter,
+)
+
+
+class InitCommand(Command):  # pylint: disable=too-few-public-methods
+    """Init Command for CLI"""
+
+    def __init__(self, cid, logger=None, **kwargs):
+        self.cid = cid
+        if logger:
```
Comment (iakov-aws) [inline cid/commands/cmd_init.py:19]: what is wrong with module level logger variable?

File: cid/commands/cmd_init.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    def _create_glue_crawler(self):
+        """Create a new Glue crawler"""
+
+    def _create_glue_database(self):
+        """Create Glue database"""
+        try:
+            self.cid.glue.create_database(name=self.database_name, catalog_id=self.catalog_id)
+            print(f"\tGlue Database...\tCreated ({self.database_name} in {self.catalog_id})")
+        except self.cid.glue.client.exceptions.AlreadyExistsException:
```
Comment (iakov-aws) [inline cid/commands/cmd_init.py:118]: good

File: cid/commands/cmd_init.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+        except Exception as ex:  # pylint: disable=broad-except
+            self._logger.exception(str(ex))
+            print("\tAthena Workgroup...\tFailed")
+
+    def _create_query_results_bucket(self):
+        """Create bucket for storing query results"""
+        try:
+            self.cid.s3.create_bucket(self.bucket_name)
+            print(f"\tAthena S3 Bucket...\tCreated ({self.bucket_name})")
+        except Exception as ex:  # pylint: disable=broad-except
```
Comment (iakov-aws) [inline cid/commands/cmd_init.py:140]: Any Exception other than "Already exists " must fail the process

File: cid/commands/cmd_init.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        )
+        counter = 0
+        while "@" not in email or "." not in email:
+            counter += 1
+            email = get_parameter("qs-notification-email", "Notification email", default=email)
+            if "@" not in email or "." not in email:
+                print(f"\t{email} does not seem to be a valid email. Please, try again.")
+                unset_parameter("qs-notification-email")
+                if counter >= 5:
+                    exit(1)
```
Comment (iakov-aws) [inline cid/commands/cmd_init.py:183]: CidCritical, not exit

File: cid/commands/cmd_init.py [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+            "Edition": "ENTERPRISE",
+            "AuthenticationMethod": "IAM_AND_QUICKSIGHT",
+            "AwsAccountId": self.cid.base.account_id,
+            "AccountName": account_name,  # Should be a parameter with a reasonable default
+            "NotificationEmail": email,  # Read the value from account parameters as a default
+        }
+
+        response = self.cid.qs.client.create_account_subscription(**params)
+        if response.get("Status", 0) == 200:
+            print("\tQuickSight Edition...\tSubscribed")
```
Comment (iakov-aws) [inline cid/commands/cmd_init.py:210]: And if not?

File: cid/helpers/account_map.py [hunk 7]
```diff
@@ -1,296 +1,281 @@
 import csv
```
Comment (iakov-aws) [inline cid/helpers/account_map.py:1]: I would suggest to revert all changes here as irrelevant.

File: cid/cli.py [hunk 8]
```diff
@@ -137,51 +140,66 @@ def delete(ctx, dashboard_id, **kwargs):
...
     """Share QuickSight resources (Dashboard, Datasets, DataSource)"""
-    
+
     ctx.obj.share(dashboard_id)
 
-if __name__ == '__main__':
+
+@click.option("-v", "--verbose", count=True)
+@click.option("-y", "--yes", help="confirm all", is_flag=True, default=False)
+@click.option("--qs-notification-email", help="QuickSight Notification Email", default="")
```
Comment (iakov-aws) [inline cid/cli.py:192]: please choose if you use click options or prameters

click options are more difficult to provide with contextual defaults

**Review by iakov-aws — CHANGES_REQUESTED** [review]

**Comment by iakov-aws** [comment]
great job. Please review the comments and fix conflicts

File: cid/commands/cmd_init.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+            assume_role_policy_document = pkg_resources.read_text("cid.builtin.core.data.iam", "glue_assume_role_policy_document.json")
+        except Exception:  # pylint: disable=broad-except
+            self._logger.exception()
+            print("\tGlue role ARN...\tERROR: Could not load role policy document")
+
+        try:
+            self.cid_crawler_role_arn = self.cid.iam.create_role(role_name=self.cid_crawler_role_name, assume_role_policy_document=assume_role_policy_document)
+            print(f"\tGlue role ARN...\tCreated ({self.cid_crawler_role_arn})")
+        except Exception:  # pylint: disable=broad-except
+            self._logger.debug("WARNING: Cannot create glue role")
```
[the remark this answers is not in the document]
Comment (gaborschulz-aws) [reply cid/commands/cmd_init.py:77]: Fixed

File: cid/utils.py [hunk 10]
```diff
@@ -177,3 +178,31 @@ def unset_parameter(param_name):
...
+    if s3_path.endswith('/'):
+        s3_path = s3_path[:-1]
+    parts = s3_path.split('/')
+    data['Bucket'] = parts[0]
+    if len(parts[1:]) == 1:  # most likely it is created by CFN or similar
+        data['Partitions'] = partitions['managed_by_cfn']
+    elif len(parts) > 3 and parts[-1] == parts[-2]:  # most likely it is manual CUR
+        data['Partitions'] = partitions['manual']
+    else:
+        raise Exception(f'CUR BucketPath={parts[0]} format is not recognized. It must be s3://(bucket)/cur or s3://{bucket}/{curprefix}/{curname}/{curname} ')  # pylint: disable=line-too-long
```
[the remark this answers is not in the document]
Comment (gaborschulz) [reply cid/utils.py:202]: Does the CidError exception already exist or should I create it from scatch?

[the remark this answers is not in the document]
Comment (iakov-aws) [reply cid/utils.py:202]: Yes it exists 

There are two exceptions for now
CidError for regular exception
CidCritical for terminating program in lambda or cli. 

For sure exceptions can be more specrec if needed.

File: cid/commands/cmd_init.py [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+                policy_document = policy_document.replace(key, value)
+
+            try:
+                policy = self.cid.iam.create_policy_from_json(policy_name=name, policy_document=policy_document)
+            except Exception:  # pylint: disable=broad-except
+                policy = f"arn:{self.aws_partition}:iam::{self.cid.base.account_id}:policy/{name}"
+            print(f"\t\t{policy}")
+            self.cid.iam.attach_policy(role_name=self.cid_crawler_role_name, policy_arn=policy)
+
+    def _create_glue_crawler(self):
```
[the remark this answers is not in the document]
Comment (gaborschulz) [reply cid/commands/cmd_init.py:113]: I'm not sure what the question is here, could you please, elaborate?

[the remark this answers is not in the document]
Comment (iakov-aws) [reply cid/commands/cmd_init.py:113]: Sure. This looks pretty empty. If this is WIP it is fine, but not for a final merge.

**Pull request closed without merging**